### PR TITLE
Add wrapped logger

### PR DESCRIPTION
**Pull request by andybradshaw**
## Before this PR

Currently, wrapped logging is not supported by the go logging libraries

## After this PR

==COMMIT_MSG==
Add support for emitting wrapped.1 logs
==COMMIT_MSG==

## Possible downsides?

In order to reduce code duplication (and potential bugs), we would have to either add the `ToParams` function to the typed logger interfaces, or export the `DefaultLogger` structs and do type assertions. I would imagine doing the later is preferable, since it doesn't really make sense to expose the params functions to the end user. 

Wanted to put this up as a draft before expanding this to other log types, since it's a decent amount of code + tests to do if we don't think this is a good approach.

---
This change is [Reviewable](https://reviewable.io/reviews/palantir/witchcraft-go-logging/119)

### CODE REVIEW

**Comment by andybradshaw** [comment]
Tests are currently failing due to how the various logger impls handle having missing/empty fields. Looking at updating the matchers accordingly, but wanted to get this up for general feedback in the meantime

**Review by andybradshaw — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/palantir/witchcraft-go-logging/119)* status: 0 of 21 files reviewed, 2 unresolved discussions

___
*[wlog/svclog/svc1log/logger_default.go, line 64 at r2](https://reviewable.io/reviews/palantir/witchcraft-go-logging/119#-MUKC8mXD91khogn22M7:-MUKC8mXD91khogn22M8:b-onzbh1) ([raw file](https://github.com/palantir/witchcraft-go-logging/blob/626e4598a9d8357462e7595fb9db97732182a47d/wlog/svclog/svc1log/logger_default.go#L64)):*
> ```Go
> }
> 
> func ToParams(msg string, level wlog.Param, inParams []Param) []wlog.Param {
> ```

Turns out this doesn't need anything from the default logger... so we can just export the function without exporting the type. Still looking into making this internal, but at least it can be separated from struct.
___
*[wlog/wrappedlog/wrapped1log/context.go, line 28 at r2](https://reviewable.io/reviews/palantir/witchcraft-go-logging/119#-MUKCWFV6CA6ihj9dnbc:-MUKCWFV6CA6ihj9dnbd:b18arqr) ([raw file](https://github.com/palantir/witchcraft-go-logging/blob/626e4598a9d8357462e7595fb9db97732182a47d/wlog/wrappedlog/wrapped1log/context.go#L28)):*
> ```Go
> // replace any logger that was previously set on the context (along with all parameters that may have been set on the
> // logger).
> func WithLogger(ctx context.Context, logger Logger) context.Context {
> ```

Pretty sure none of this is necessary, since a user will never actually store a wrapped logger, just the other loggers that the wrapped logger implements... will remove.

File: wlog/diaglog/diag1log/logger_default.go [hunk 1]
```diff
@@ -28,17 +30,18 @@ func (l *defaultLogger) Diagnostic(diagnostic logging.Diagnostic, params ...Para
...
+	outParams[len(defaultParams)] = wlog.NewParam(diagnosticParam(diagnostic).apply)
 	for idx := range inParams {
-		outParams[len(defaultTypeParam)+1+idx] = wlog.NewParam(inParams[idx].apply)
+		outParams[len(defaultParams)+1+idx] = wlog.NewParam(inParams[idx].apply)
 	}
 	return outParams
 }
 
-var defaultTypeParam = []wlog.Param{
+var defaultParams = []wlog.Param{
```
Comment (nmiyake) [inline wlog/diaglog/diag1log/logger_default.go]: Is there a reason that this is being renamed?

File: wlog/svclog/svc1log/context.go [hunk 2]
```diff
@@ -51,7 +51,7 @@ func WithLoggerParams(ctx context.Context, params ...Param) context.Context {
 }
 
 // FromContext returns the Logger stored in the provided context. If no logger is set on the context, returns the logger
-// created by calling DefaultLogger. If the context contains a TraceID set using wtracing, the returned logger has that
+// created by calling defaultLogger. If the context contains a TraceID set using wtracing, the returned logger has that
```
Comment (nmiyake) [inline wlog/svclog/svc1log/context.go]: Extraneous edit caused by refactor?

File: wlog/svclog/svc1log/logger_default.go [hunk 3]
```diff
@@ -15,9 +15,24 @@
 package svc1log
 
 import (
+	"time"
+
 	"github.com/palantir/witchcraft-go-logging/wlog"
 )
 
+var DebugLevelParam = wlog.NewParam(func(entry wlog.LogEntry) {
```
Comment (nmiyake) [inline wlog/svclog/svc1log/logger_default.go]: What's the high-level motivation/explanation for why the logger level is now a param?

If we were to actually do this, then these should probably be functions that return the param -- it's not good practice to have exported package-level variables since they are mutable (anyone importing the package could do something like `svc1log.DebugLevelParam = nil` and reassign the global value)

File: wlog/wrappedlog/wrapped1log/context.go [hunk 4]
```diff
@@ -0,0 +1,31 @@
+// Copyright (c) 2018 Palantir Technologies. All rights reserved.
```
Comment (nmiyake) [inline wlog/wrappedlog/wrapped1log/context.go]: nit: since this is a new file, copyright should be 2021 (I'm guessing this was just copy-pasted -- if you omit the header entirely, then the "verify" task will add the proper header with the current year as part of the "license" task)

File: wlog/wrappedlog/wrapped1log/context.go [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+
+type wrapped1LogContextKeyType string
+
+const contextKey = wrapped1LogContextKeyType(TypeValue)
+
+// WithLogger returns a copy of the provided context with the provided Logger included as a value. This operation will
+// replace any logger that was previously set on the context (along with all parameters that may have been set on the
+// logger).
+func WithLogger(ctx context.Context, logger Logger) context.Context {
+	// Set _all_ the underlying logger types
```
Comment (nmiyake) [inline wlog/wrappedlog/wrapped1log/context.go]: What does this comment mean? Does this code do something beyond just replace the logger on the specific context?

File: wlog/logger_provider_jsonmarshal.go [hunk 6]
```diff
@@ -41,28 +41,34 @@ func (l *jsonMapLogger) Log(params ...Param) {
...
+		ReverseParams(params)
+		l.logOutput(params)
 	}
 }
 
 func (l *jsonMapLogger) Info(msg string, params ...Param) {
 	switch l.level {
 	case DebugLevel, InfoLevel:
-		l.logOutput(append(params, StringParam("message", msg), StringParam("level", "INFO")))
+		l.logOutput(append([]Param{StringParam("message", msg), StringParam("level", "INFO")}, params...))
```
Comment (nmiyake) [inline wlog/logger_provider_jsonmarshal.go]: Why does this logic for "Info" not match the others?

File: wlog/wrappedlog/wrapped1log/context.go (hunk 5 above)
Comment (nmiyake) [reply wlog/wrappedlog/wrapped1log/context.go]: Never mind, see updated comment that this will not be needed

**Review by nmiyake — COMMENTED** [review]
Thanks for the PR -- read through after our talk earlier today and have a much better understanding of the goal and the proposed implementation here.

Did you consider an alternate approach where each individual logger type supports instantiating a "wrapped" version? Was wondering that because, with the current approach, the "wrapped" logging package basically needs to have intimate knowledge of how each logger type it wraps operates and also sometimes requires the logger being wrapped to export functions to enable this.

Having the wrapped logger be its own package would make more sense if it could generically wrap any type of logger, but if we have to custom-implement the wrapped logger per logger type anyway then I'm not sure that it actually provides an advantage. In the alternate approach, the `wrapped1log` would basically just export the constants for things like the type value, payload etc.

May make sense to schedule another time for us to sync up now that I feel more familiar with the problem space and the proposed implementation.

File: wlog/logger_provider_jsonmarshal.go (hunk 6 above)
Comment (advayakrishna) [reply wlog/logger_provider_jsonmarshal.go]: Fixed, I think this is the correct logic

File: wlog/diaglog/diag1log/logger_default.go [hunk 7]
```diff
@@ -28,17 +30,18 @@ func (l *defaultLogger) Diagnostic(diagnostic logging.Diagnostic, params ...Para
...
+	outParams[len(defaultTypeParams)] = wlog.NewParam(diagnosticParam(diagnostic).apply)
 	for idx := range inParams {
-		outParams[len(defaultTypeParam)+1+idx] = wlog.NewParam(inParams[idx].apply)
+		outParams[len(defaultTypeParams)+1+idx] = wlog.NewParam(inParams[idx].apply)
 	}
 	return outParams
 }
 
-var defaultTypeParam = []wlog.Param{
+var defaultTypeParams = []wlog.Param{
```
Comment (nmiyake) [inline wlog/diaglog/diag1log/logger_default.go]: Name should be kept as `defaultTypeParam` -- there's still only one param entry in the slice, and don't think that the change is significant enough to rename and introduce the new diff.

File: wlog/logger_provider_jsonmarshal.go [hunk 8]
```diff
@@ -41,28 +41,28 @@ func (l *jsonMapLogger) Log(params ...Param) {
 func (l *jsonMapLogger) Debug(msg string, params ...Param) {
 	switch l.level {
 	case DebugLevel:
-		l.logOutput(append(params, StringParam("message", msg), StringParam("level", "DEBUG")))
+		l.logOutput(append([]Param{StringParam("message", msg), StringParam("level", "DEBUG")}, params...))
```
Comment (nmiyake) [inline wlog/logger_provider_jsonmarshal.go]: Why are params at the end now rather than the beginning? Not a big deal as long as everything works, but would like to understand if this is intentional/necessary and, if so, why it works.

File: wlog/svclog/svc1log/context.go [hunk 9]
```diff
@@ -84,7 +84,7 @@ func safeAndUnsafeParamsFromParams(params []Param) (safe map[string]interface{},
 }
 
 // loggerFromContext returns the logger stored in the provided context. If no logger is set on the context, returns the
-// logger created by calling DefaultLogger.
+// logger created by calling defaultLogger.
```
Comment (nmiyake) [inline wlog/svclog/svc1log/context.go]: nit: should be `defaultLoggerCreator`.

Searching through the codebase, it looks like this typo/incorrect attribution happens for most of the logger types. Given that this PR doesn't modify this function, I would prefer that this not be corrected in this PR and we open a separate PR that fixes this documentation issue for all the relevant functions.

File: wlog/svclog/svc1log/logger_default.go [hunk 10]
```diff
@@ -15,9 +15,37 @@
 package svc1log
 
 import (
+	"time"
+
 	"github.com/palantir/witchcraft-go-logging/wlog"
 )
 
+var debugLevelParam = wlog.NewParam(func(entry wlog.LogEntry) {
```
Comment (nmiyake) [inline wlog/svclog/svc1log/logger_default.go]: Is the declaration of these variables as top-level variables rather than inlining as part of the function return intentional for purposes of performance/allocation? If so, then that's fine, but would add comment that explains this and then declare these in block form:

```go
var (
   debugLevelParam = ...
   infoLevelParam = ...
   // etc.
)
```

If there isn't a performance-related or otherwise intentional reason for these declarations, then would prefer that they not be declared as variables and inline in the functions instead.

File: wlog/svclog/svc1log/logger_default.go [hunk 11]
```diff
@@ -26,39 +54,42 @@ type defaultLogger struct {
...
+	})
 	for idx := range inParams {
-		outParams[len(defaultTypeParam)+idx] = wlog.NewParam(inParams[idx].apply)
+		outParams[len(defaultParams)+2+idx] = wlog.NewParam(inParams[idx].apply)
 	}
 	return outParams
 }
 
-var defaultTypeParam = []wlog.Param{
+var defaultParams = []wlog.Param{
```
Comment (nmiyake) [inline wlog/svclog/svc1log/logger_default.go]: Do not rename variable

File: wlog-zap/internal/provider.go [hunk 12]
```diff
@@ -33,7 +32,6 @@ type loggerProvider struct{}
 
 func (lp *loggerProvider) NewLogger(w io.Writer) wlog.Logger {
 	logger, atomicLevel := newZapLogger(w, wlog.InfoLevel, zapcore.EncoderConfig{
-		TimeKey:        wlog.TimeKey,
```
Comment (nmiyake) [inline wlog-zap/internal/provider.go]: Would prefer that this refactor be done as a separate commit that doesn't add wrapped logger

**Review by nmiyake — CHANGES_REQUESTED** [review]
Thanks for updating! Took a deeper look.

On the whole, am on board with this change. There are some minor points of feedback that I left as part of the review.

At a macro level, would request that the change that makes the logger provider API more generic (removes `TimeKey` and `LevelKey` etc. from the generic logger providers) be made as a separate PR -- this is a behavioral change to publicly exported API, so would like to have it be reviewed and documented separately.

File: wlog/logger_provider_jsonmarshal.go (hunk 8 above)
Comment (advayakrishna) [reply wlog/logger_provider_jsonmarshal.go]: Yeah, this is intentional so that the svc1logger can overwrite the message field using a param, want its params to be applied after the ones defined here.

File: wlog/svclog/svc1log/logger_default.go (hunk 10 above)
Comment (advayakrishna) [reply wlog/svclog/svc1log/logger_default.go]: Yeah, this is so they are initialized once. Will update.

File: wlog-zap/internal/provider.go (hunk 12 above)
Comment (advayakrishna) [reply wlog-zap/internal/provider.go]: Done, opened https://github.com/palantir/witchcraft-go-logging/pull/123/files

File: wlog/wrappedlog/wrapped1log/context.go (hunk 5 above)
Comment (nmiyake) [reply wlog/wrappedlog/wrapped1log/context.go]: Is this still needed? Per my last comment, thought you had a previous comment that stated that this wouldn't need to be set here

File: wlog/wrappedlog/wrapped1log/logger.go [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+
+	"github.com/palantir/witchcraft-go-logging/wlog"
+	"github.com/palantir/witchcraft-go-logging/wlog/svclog/svc1log"
+)
+
+type Logger interface {
+	Service(params ...svc1log.Param) svc1log.Logger
+}
+
+func New(w io.Writer, level wlog.LogLevel, name string, version string) Logger {
```
Comment (nmiyake) [inline wlog/wrappedlog/wrapped1log/logger.go]: nit: combine declaration for last two parameters:

```
name, version string)
```

File: wlog/wrappedlog/wrapped1log/logger.go [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+	return NewFromProvider(w, level, wlog.DefaultLoggerProvider(), name, version)
+}
+
+func NewFromProvider(w io.Writer, level wlog.LogLevel, creator wlog.LoggerProvider, name string, version string) Logger {
```
Comment (nmiyake) [inline wlog/wrappedlog/wrapped1log/logger.go]: Same as above

File: wlog/wrappedlog/wrapped1log/logger_service.go [hunk 15]
```diff
@@ -0,0 +1,56 @@
...
+
+func (l *wrappedSvc1Logger) Error(msg string, params ...svc1log.Param) {
+	l.logger.Error("", l.ToServiceParams(msg, svc1log.ErrorLevelParam(), params)...)
+}
+
+func (l *wrappedSvc1Logger) SetLevel(level wlog.LogLevel) {
+	l.logger.SetLevel(level)
+}
+
+func (l *wrappedSvc1Logger) ToServiceParams(message string, levelParam wlog.Param, inParams []svc1log.Param) []wlog.Param {
```
Comment (nmiyake) [inline wlog/wrappedlog/wrapped1log/logger_service.go]: Does this function need to be exported? If not, would keep internal.

**Review by nmiyake — CHANGES_REQUESTED** [review]
Looks pretty good overall! Some feedback.

File: changelog/@unreleased/pr-119.v2.yml [hunk 16]
```diff
@@ -0,0 +1,8 @@
+type: improvement
```
Comment (nmiyake) [inline changelog/@unreleased/pr-119.v2.yml]: Can you make this (and type field below) "feature"

**Review by nmiyake — CHANGES_REQUESTED** [review]

File: wlog/wrappedlog/wrapped1log/logger_service.go [hunk 17]
```diff
@@ -0,0 +1,56 @@
+// Copyright (c) 2018 Palantir Technologies. All rights reserved.
```
Comment (nmiyake) [inline wlog/wrappedlog/wrapped1log/logger_service.go]: Oops this should be 2021

**Review by nmiyake — APPROVED** [review]
Thanks!
